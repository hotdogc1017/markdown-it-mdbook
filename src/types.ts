export interface BookOptions {
  book: {
    src: string;
    title?: string;
    authors?: string[];
    description?: string;
    language?: string;
    /**
     * unsupported
     */
    "text-direction"?: "ltr" | "rtl";
  };
  /**
   * unsupported
   */
  rust?: {
    edition?: "2015" | "2018" | "2021";
  };
  /**
   * unsupported
   */
  build?: {
    "build-dir"?: string;
    "create-missing"?: boolean;
    "use-default-preprocessors"?: boolean;
    "extra-watch-dirs"?: string[];
  };
  /**
   * unsupported
   */
  preprocessor?: {
    [key: string]: any;
  };
  output?: {
    html?: {
      code: {
        hidelines: Record<string, string>;
      };
      /**
       * unsupported
       */
      "additional-css"?: string[];
      /**
       * unsupported
       */
      search?: {
        "limit-results"?: number;
      };
    };
  };
}
