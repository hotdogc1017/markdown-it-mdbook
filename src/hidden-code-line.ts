import MarkdownIt from "markdown-it";
import type { BookOptions } from "./types";

/**
 * prefix priority: local > global > default
 */
export function hiddenCodeLinePlugin(
  md: MarkdownIt,
  bookOptions?: BookOptions,
) {
  const globalPrefixMap: Record<string, string> = {
    rust: "#",
    ...getGloabalPrefixMap(bookOptions),
  };
  const originalRender = md.renderer.rules.fence!;
  md.renderer.rules.fence = (...args) => {
    const [tokens, idx] = args;
    const token = tokens[idx];
    const { lang, hidelines: localPrefix } = parseInfo(token.info);
    const globalPrefix = globalPrefixMap[lang];

    if (!localPrefix && !globalPrefix) {
      return originalRender(...args);
    }

    const prefix = localPrefix || globalPrefix;
    const RE = new RegExp(`^${prefix}.*`);
    const lines = token.content
      .split("\n")
      .filter((line) => !RE.test(line.trim()));
    token.info = `${lang}`;
    token.content = lines.join("\n");
    return originalRender(...args);
  };
}

export function parseInfo(info: string) {
  const [lang, ...attrs] = info.trim().split(",");
  const hidelinesStr = attrs.find((attr) => {
    const [key] = attr.split("=");
    return key.trim() === "hidelines";
  });

  if (!hidelinesStr) {
    return { lang };
  }

  const [, hidelines] = hidelinesStr.split("=");
  return { lang, hidelines };
}

function getGloabalPrefixMap(bookOptions?: BookOptions) {
  return bookOptions?.output?.html?.code.hidelines ?? {};
}
