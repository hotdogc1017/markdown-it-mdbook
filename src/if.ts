import {
  createBlockRuleParser,
  type ParserArgs,
  type RenderArgs,
  type TokenWithMeta,
} from "./utils";
import MarkdownIt from "markdown-it";
import Token from "markdown-it/lib/token.mjs";
import type { RuleBlock } from "markdown-it/lib/parser_block.mjs";
import path from "node:path";
import fs from "node:fs";

/**
 * Rust needs to be supported for now.
 *
 * Captures: [anchor_tag, anchor]
 */
const ANCHOR_RE_LIST = [
  /^\/\/+ #?(ANCHOR(?:_END)?): ([a-zA-Z]+)$/, // javascript, typescript, java, Rust
  /^\/\* ?#(ANCHOR(?:_END)?): ([a-zA-Z]+) ?\*\/$/, // css, less, scss
  /^#pragma (ANCHOR(?:_END)?): ([a-zA-Z]+)$/, // C, C++
  /^<!-- #?(ANCHOR(?:_END)?): ([a-zA-Z]+) -->$/, // HTML, markdown
  /^#(ANCHOR(?:_END )): ([a-zA-Z]+)$/, // Visual Basic
  /^::#(ANCHOR(?:_END)): ([a-zA-Z]+)$/, // Bat
  /^# ?(ANCHOR(?:_END)?): ([a-zA-Z]+)$/, // C#, PHP, Powershell, Python, perl & misc
  /^\/\* (ANCHOR(?:_END)?): ([a-zA-Z]+) \*\/$/, // Rust
];

/**
 * Captures: [filepath, lang, anchor, startLine, endLine]
 */
export const INCLUDE_RE =
  /^\s*\{{2}#include\s([^<>|:"*?]+(?:\.([a-z0-9]+)))(?::?([a-zA-Z]+))?(:\d*)?(:\d*)?\}{2}\s*$/;

export function includingFilesPlugin(md: MarkdownIt) {
  type MatchResult = ReturnType<typeof syntaxMatcher>;
  const ruleName = "fence";
  const parser = (...args: ParserArgs) => {
    return createBlockRuleParser({
      ruleName,
      syntaxMatcher,
      originalArgs: args,
      setToken: (token: TokenWithMeta<MatchResult>, match: MatchResult) => {
        token.markup = "```";
        // TODO: Compatible vitepress
        token.info = "";
        token.content = match.filepath;
      },
    });
  };

  const render = (...args: RenderArgs<MatchResult, { cwd: string }>) => {
    const [tokens, idx, , env] = args;
    const { src } = tokens[idx];
    if (!env?.cwd || path.isAbsolute(env.cwd)) {
      console.warn(
        `[markdwon-it-mdBook] ${env.cwd} is not a valid path or absolute path`,
      );
      return `<p>[markdwon-it-mdBook] unable to resolve file: ${src}</p>`;
    }
    const fileContent = fs.readFileSync(path.resolve(env.cwd, src.filepath), {
      encoding: "utf-8",
    });

    const renderWithFence = () => {
      for (const token of tokens) {
        token.type = "fence";
        token.info = src.lang || "text";
        token.markup = "```";
        token.content = fileContent;
      }
    };

    return;
  };

  md.block.ruler.before("paragraph", ruleName, parser);
  md.renderer.rules[ruleName] = render;
}

export function syntaxMatcher(line: string) {
  let [, filepath, lang, anchor, startLine, endLine] =
    INCLUDE_RE.exec(line) || [];

  if (!filepath) return null;

  // The lines between startLine and endLine will be excluded
  let excludeMode = false;
  if (!anchor) {
    if (startLine && startLine !== ":" && endLine === ":") {
      const start = parseInt(startLine.slice(1));
      startLine = ":";
      endLine = `:${start - 1}`;
      excludeMode = true;
    }
    startLine = startLine?.slice(1) || "0";
    // if endLine is undefined, it defaults to startLine, which will only import one line
    endLine = endLine?.slice(1) || startLine;
  }

  return {
    filepath,
    lang,
    anchor,
    startLine,
    endLine,
    excludeMode,
  };
}

export function findLinesWithAnchor(lines: string[], anchor: string) {
  let RE = null;
  const pos = { start: -1, end: -1 };

  lines.forEach((line, i) => {
    if (!RE) {
      const anchorRE = ANCHOR_RE_LIST.find((re) => re.test(line));
      const [, anchor_tag, anchor_name] = anchorRE?.exec(line) ?? [];
      if (!anchor_tag || anchor_name !== anchor) {
        return;
      } else {
        RE = anchorRE;
      }
    }

    const [_, tag, name] = RE.exec(line) ?? [];
    if (!tag || name !== anchor) {
      return;
    }

    if (!tag.endsWith("_END")) {
      pos.start = i + 1;
    } else if (tag.endsWith("_END") && pos.start >= 0) {
      pos.end = i;
    }
  });

  if (pos.start < 0 || pos.end < 0) {
    return [];
  }

  return lines.slice(pos.start, pos.end);
}

export function findLines(
  lines: string[],
  start: number | string,
  end: number | string,
  excludeMode?: boolean,
) {
  const startLine = parseInt(start.toString());
  const endLine = parseInt(end.toString());

  if (startLine < 0 || endLine < 0) {
    throw [];
  }

  if (excludeMode) {
    return [...lines.slice(0, startLine), ...lines.slice(endLine + 1)];
  }

  return lines.slice(startLine, endLine + 1);
}
