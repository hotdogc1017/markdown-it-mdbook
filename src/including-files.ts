import {
  createBlockRuleParser,
  type ParserArgs,
  type RenderArgs,
  type TokenWithMeta,
} from "./utils";
import MarkdownIt from "markdown-it";
import Token from "markdown-it/lib/token.mjs";
import path from "node:path";
import fs from "node:fs";

/**
 * Rust needs to be supported for now.
 *
 * Captures: [anchor_tag, anchor]
 */
const ANCHOR_RE_LIST = [
  /^\/\/+ #?(ANCHOR(?:_END)?): ([a-zA-Z]+)$/, // javascript, typescript, java, Rust
  /^\/\* ?#(ANCHOR(?:_END)?): ([a-zA-Z]+) ?\*\/$/, // css, less, scss
  /^#pragma (ANCHOR(?:_END)?): ([a-zA-Z]+)$/, // C, C++
  /^<!-- #?(ANCHOR(?:_END)?): ([a-zA-Z]+) -->$/, // HTML, markdown
  /^#(ANCHOR(?:_END )): ([a-zA-Z]+)$/, // Visual Basic
  /^::#(ANCHOR(?:_END)): ([a-zA-Z]+)$/, // Bat
  /^# ?(ANCHOR(?:_END)?): ([a-zA-Z]+)$/, // C#, PHP, Powershell, Python, perl & misc
  /^\/\* (ANCHOR(?:_END)?): ([a-zA-Z]+) \*\/$/, // Rust
];

/**
 * Captures: [filepath, lang, anchor, startLine, endLine]
 */
export const INCLUDE_RE =
  /^\s*\{{2}#include\s([^<>|:"*?]+(?:\.([a-z0-9]+)))(?::?([a-zA-Z]+))?(:\d*)?(:\d*)?\}{2}\s*$/;

export function includingFilesPlugin(md: MarkdownIt) {
  type MatchResult = ReturnType<typeof syntaxMatcher>;
  const ruleName = "fence";

  const setToken = (token: TokenWithMeta<MatchResult>, match: MatchResult) => {
    token.markup = "```";
    // TODO: Compatible vitepress
    token.info = `${match.lang}`;
    token.content = match.filepath;
  };

  const parser = (...args: ParserArgs) => {
    return createBlockRuleParser({
      ruleName,
      syntaxMatcher,
      originalArgs: args,
      setToken,
    });
  };

  const fenceRender = md.renderer.rules["fence"]!;
  const render = (...args: RenderArgs<MatchResult, { cwd: string }>) => {
    const [tokens, idx, , env] = args;
    const token = tokens[idx];
    const { src } = token;

    if (!src) {
      // try to parse again, if successful, set token and recursion
      const result = syntaxMatcher(token.content);
      if (result) {
        setToken(token, result);
        token.src = result;
        return render(...args);
      } else {
        return fenceRender(...args);
      }
    }

    if (!env?.cwd || !path.isAbsolute(env.cwd)) {
      console.warn(
        `[markdwon-it-mdBook] ${env.cwd} is not a valid path or absolute path`,
      );
      token.content = `[markdwon-it-mdBook] unresolved file: ${src.filepath}`;
      return fenceRender(...args);
    }

    const stat = fs.statSync(env.cwd);
    const filepath = path.join(
      stat.isDirectory() ? env.cwd : path.dirname(env.cwd),
      `./${src.filepath}`,
    );
    if (!fs.existsSync(filepath)) {
      // print absolute path for debugging
      console.warn(`[markdwon-it-mdBook] ${filepath} is not found`);
      token.content = `${src.filepath} is not found`;
      return fenceRender(...args);
    } else if (fs.statSync(filepath).isDirectory()) {
      console.warn(
        `[markdwon-it-mdBook] require file, but get a directory: ${src.filepath}`,
      );
      token.content = `require file, but get a directory: ${src.filepath}`;
      return fenceRender(...args);
    }
    const fileContent = fs.readFileSync(filepath, {
      encoding: "utf-8",
    });

    if (!src.anchor && !src.startLine && !src.endLine) {
      token.content = fileContent;
      return fenceRender(...args);
    }

    const includedLines = src.anchor
      ? findLinesWithAnchor(fileContent.split("\n"), src.anchor)
      : findLines(
          fileContent.split("\n"),
          src.startLine,
          src.endLine,
          src.excludeMode,
        );

    token.content = includedLines.join("\n");

    return fenceRender(...args);
  };

  md.block.ruler.before("paragraph", ruleName, parser);
  md.renderer.rules[ruleName] = render;
}

export function syntaxMatcher(line: string) {
  let [, filepath, lang, anchor, startLine, endLine] =
    INCLUDE_RE.exec(line.trim()) || [];

  if (!filepath) return null;

  // The lines between startLine and endLine will be excluded
  let excludeMode = false;
  if (!anchor) {
    if (startLine && startLine !== ":" && endLine === ":") {
      const start = parseInt(startLine.slice(1));
      startLine = ":";
      endLine = `:${start - 1}`;
      excludeMode = true;
    }
    startLine = startLine?.slice(1) || "0";
    // if endLine is undefined, it defaults to startLine, which will only import one line
    endLine = endLine?.slice(1) || startLine;
  }

  return {
    filepath,
    lang,
    anchor,
    startLine,
    endLine,
    excludeMode,
  };
}

export function findLinesWithAnchor(lines: string[], anchor: string) {
  let RE = null;
  const pos = { start: -1, end: -1 };

  lines.forEach((line, i) => {
    if (!RE) {
      const anchorRE = ANCHOR_RE_LIST.find((re) => re.test(line));
      const [, anchor_tag, anchor_name] = anchorRE?.exec(line) ?? [];
      if (!anchor_tag || anchor_name !== anchor) {
        return;
      } else {
        RE = anchorRE;
      }
    }

    const [_, tag, name] = RE.exec(line) ?? [];
    if (!tag || name !== anchor) {
      return;
    }

    if (!tag.endsWith("_END")) {
      pos.start = i + 1;
    } else if (tag.endsWith("_END") && pos.start >= 0) {
      pos.end = i;
    }
  });

  if (pos.start < 0 || pos.end < 0) {
    return [];
  }

  return lines.slice(pos.start, pos.end);
}

export function findLines(
  lines: string[],
  start: number | string,
  end: number | string,
  excludeMode?: boolean,
) {
  const startLine = parseInt(start.toString());
  const endLine = parseInt(end.toString());

  if (startLine < 0 || endLine < 0) {
    throw [];
  }

  if (excludeMode) {
    return [...lines.slice(0, startLine), ...lines.slice(endLine + 1)];
  }

  return lines.slice(startLine, endLine + 1);
}
