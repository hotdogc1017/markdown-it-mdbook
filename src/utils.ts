import Renderer from "markdown-it/lib/renderer.mjs";
import StateBlock from "markdown-it/lib/rules_block/state_block.mjs";
import Token from "markdown-it/lib/token.mjs";

interface CreateBlockRuleParserOptions {
  ruleName: string;
  syntaxMatcher: (line: string) => unknown | null;
  originalArgs: ParserArgs;
  setToken?: (token: TokenWithMeta, match: unknown) => void;
}

export type ParserArgs = [StateBlock, number, number, boolean];
export type RenderArgs<T = any, E = any> = [
  TokenWithMeta<T>[],
  number,
  any,
  E,
  Renderer,
];

/**
 * Specify the type of meta in the token for content
 */
export type TokenWithMeta<T = any> = Token & { src: T };

/**
 * !!!NOTE: 仅适合创建不需要跨行解析的语法
 *
 * This func helps to quickly create a block rule and handles same edge cases automatically:
 * - When indent more than 3 spaces, it should be a code block.
 * - Skip if silent mode
 * - Auto increment line number
 * - Simplify params passing
 * - Manage rule name
 */
export function createBlockRuleParser({
  ruleName,
  syntaxMatcher,
  originalArgs,
  setToken,
}: CreateBlockRuleParserOptions) {
  const [state, startLine, endLine, silent] = originalArgs;
  const pos = state.bMarks[startLine] + state.tShift[startLine];
  const max = state.eMarks[startLine];

  if (state.sCount[startLine] - state.blkIndent >= 4) {
    return false;
  }
  if (silent) {
    return true;
  }
  const line = state.src.slice(pos, max);
  const match = syntaxMatcher?.(line);
  if (!match) {
    return false;
  }
  const token = state.push(ruleName, "", 1) as TokenWithMeta;
  token.map = [startLine, startLine + 1];
  token.src = match;
  setToken?.(token, match);

  state.line = startLine + 1;
  return true;
}
