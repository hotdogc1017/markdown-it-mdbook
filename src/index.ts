import { hiddenCodeLinePlugin } from "./hidden-code-line";
import { includingFilesPlugin } from "./including-files";
import type { BookOptions } from "./types";
import MarkdownIt from "markdown-it";

export interface MdBookPluginOptions {
  bookOptions: BookOptions;
}

export function mdBookPlugin(md: MarkdownIt, options?: MdBookPluginOptions) {
  const { bookOptions } = options || {};
  md.use(hiddenCodeLinePlugin, bookOptions);
  md.use(includingFilesPlugin);
}
