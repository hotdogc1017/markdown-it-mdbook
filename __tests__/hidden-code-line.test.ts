import { test, expect, describe } from "vitest";
import { parseInfo, hiddenCodeLinePlugin } from "../src/hidden-code-line";
import path from "node:path";
import fs from "node:fs";
import MarkdownIt from "markdown-it";

describe("parseInfo", () => {
  test("test parseInfo", () => {
    expect(parseInfo("python,hidelines=!!!")).toMatchSnapshot();
  });

  test("test parseInfo with playground", () => {
    expect(parseInfo("python,hidelines=!!!,playground")).toMatchSnapshot();
  });

  test("test parseInfo with playground and hidelines", () => {
    expect(parseInfo("python,playground,hidelines=!!!")).toMatchSnapshot();
  });

  test("test parseInfo with nothing", () => {
    expect(parseInfo("python")).toMatchSnapshot();
  });
});

describe("no-hidden-code-line plugin", () => {
  test("test no-hidden-code-line plugin", () => {
    const filesPath = path.resolve(import.meta.dirname, "./fixtures");
    const hiddenCodeLineMd = fs.readFileSync(
      path.join(filesPath, "hidden-code-line.md"),
      { encoding: "utf-8" },
    );
    const noHiddenCodeLineMd = fs.readFileSync(
      path.join(filesPath, "no-hidden-code-line.md"),
      { encoding: "utf-8" },
    );
    const pluginRenderOutput = () => {
      const bookOptions = {
        output: {
          html: {
            code: {
              hidelines: { python: "~" },
            },
          },
        },
      };
      return new MarkdownIt()
        .use(hiddenCodeLinePlugin, bookOptions)
        .render(hiddenCodeLineMd);
    };
    const normalRenderOutput = () => {
      return new MarkdownIt().render(noHiddenCodeLineMd);
    };
    expect(pluginRenderOutput()).toBe(normalRenderOutput());
  });
});
