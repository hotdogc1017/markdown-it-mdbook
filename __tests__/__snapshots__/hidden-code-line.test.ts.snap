// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseInfo > test parseInfo 1`] = `
{
  "hidelines": "!!!",
  "lang": "python",
}
`;

exports[`parseInfo > test parseInfo with nothing 1`] = `
{
  "lang": "python",
}
`;

exports[`parseInfo > test parseInfo with playground 1`] = `
{
  "hidelines": "!!!",
  "lang": "python",
}
`;

exports[`parseInfo > test parseInfo with playground and hidelines 1`] = `
{
  "hidelines": "!!!",
  "lang": "python",
}
`;
